### PR TITLE
kernel: convert static_buf!() to be like singleton

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

This PR changes `static_buf!()` to now panic if the same static buffer is used multiple times. This happens if the same `static_buf!()` expansion is run multiple times, such is if `static_buf!()` is in a function which is called multiple times. Since doing this causes memory to be aliased which is bad, we panic to let the board author know this needs to be fixed.

Implements #2564.

### Testing Strategy

I added

```rust
    let crc2 = components::crc::CrcComponent::new(
        board_kernel,
        capsules::crc::DRIVER_NUM,
        &peripherals.crccu,
    )
    .finalize(components::crc_component_helper!(sam4l::crccu::Crccu));
```
to the end of hail's main.rs, and when running I get:

```
[INFO   ]  ----- Waiting for serial port to reconnect...

panicked at 'ERROR! The same static_buf!() cannot be used multiple times.', boards/components/src/crc.rs:63:23
	Kernel version release-2.1-76-g4f51e0ed5

---| No debug queue found. You can set it with the DebugQueue component.

---| Cortex-M Fault Status |---
No Cortex-M faults detected.

---| App Status |---
```

### TODO or Help Wanted

Thoughts? Comments?

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Comment by hudson-ayers** [comment]
I pushed a commit that brings the size impact of this change on Imix from 4688 bytes to 1920 bytes. This is still a pretty big size hit, even though it is likely < 20 bytes per call to `static_init!()`.

**Comment by alexandruradovici** [comment]
What about having the check only for debug builds? I think this is somehow consistent with Rust's approach to number overflowing.

**Comment by hudson-ayers** [comment]
> What about having the check only for debug builds? I think this is somehow consistent with Rust's approach to number overflowing.

With this check, `static_init!()` will not have to be `unsafe`, which is very valuable in reducing the total amount of `unsafe` in the kernel. However if the checks only apply for debug builds we cannot remove `unsafe` since misuse can lead to UB. Also, Tock builds are `release` by default, so no one ever uses debug builds in practice.

**Comment by bradjc** [comment]
I added a commit updating the `Soundness.md` documentation. Please take a look that what I added is correct.

File: doc/Soundness.md [hunk 1]
```diff
@@ -63,10 +63,10 @@ However, in a few cases, it is useful to start with a mutable reference in order
...
 SPI driver is, for practical reasons, deferred until after construction but we
 would like to enforce that it can only be called by the platform initialization
-function (before `main` starts). This is enforced because all references after
-the platform is setup are immutable, and the `config_buffers` method takes an
-`&mut self` (_Note: it looks like this is not strictly necessary, so maybe not a
-big deal if we can't do this_).
+function (before the kernel loop starts). This is enforced because all
+references after the platform is setup are immutable, and the `config_buffers`
+method takes an `&mut self` (_Note: it looks like this is not strictly
+necessary, so maybe not a big deal if we can't do this_).
```
Comment (hudson-ayers) [inline doc/Soundness.md:69]: IMO, we could just remove this paragraph. The SPI buffers are held in `TakeCell`s now anyway, and using `&mut self` as a way to enforce that buffers are only configured at boot is not a pattern we have really carried over to other capsules.

**Review by hudson-ayers — APPROVED** [review]

**Comment by bradjc** [comment]
Closed in favor of https://github.com/tock/tock/pull/3239

**Pull request closed without merging**